lib/AST/ClangTypeConverter.cpp

```diff
@@ -118,23 +118,34 @@ const clang::Type *ClangTypeConverter::getFunctionType(
     ArrayRef<AnyFunctionType::Param> params, Type resultTy,
     AnyFunctionType::Representation repr) {
 
-  SmallVector<Type, 4> paramsTy;
-  for (auto p : params)
-    paramsTy.push_back(p.getPlainType());
+#define HANDLE_NULL_AND_FUNCTION_TYPE(ty)            \
+  if (ty.isNull())                                   \
+    return nullptr;                                  \
+  if (auto *fnTy = ty->getAs<clang::FunctionType>()) \
+    ty = ClangASTContext.getPointerType(ty);
 
   auto resultClangTy = convert(resultTy);
-  if (resultClangTy.isNull())
-    return nullptr;
-
-  SmallVector<clang::QualType, 8> paramsClangTy;
-  for (auto p : paramsTy) {
-    auto pc = convert(p);
-    if (pc.isNull())
-      return nullptr;
+  HANDLE_NULL_AND_FUNCTION_TYPE(resultClangTy);
+
+  SmallVector<clang::FunctionProtoType::ExtParameterInfo, 4> extParamInfos;
+  SmallVector<clang::QualType, 4> paramsClangTy;
+  bool someParamIsConsumed = false;
+  for (auto p : params) {
+    auto pc = convert(p.getPlainType());
+    HANDLE_NULL_AND_FUNCTION_TYPE(pc);
+    clang::FunctionProtoType::ExtParameterInfo extParamInfo;
+    if (p.getParameterFlags().isOwned()) {
+      someParamIsConsumed = true;
+      extParamInfo = extParamInfo.withIsConsumed(true);
+    }
+    extParamInfos.push_back(extParamInfo);
     paramsClangTy.push_back(pc);
   }
+#undef HANDLE_NULL_AND_FUNCTION_TYPE
 
   clang::FunctionProtoType::ExtProtoInfo info(clang::CallingConv::CC_C);
+  if (someParamIsConsumed)
+    info.ExtParameterInfos = extParamInfos.begin();
```

Comment:
This is the right way to handle this given the current Clang behavior, although I agree that it would be better for Clang to do this canonicalization itself when uniquing.